cargo-dist/src/tasks.rs

```diff
@@ -2428,17 +2434,112 @@ pub(crate) struct AnnouncementTag {
     pub rust_releases: Vec<(PackageIdx, Vec<String>)>,
 }
 
+/// details on what we're announcing (partially computed)
+struct PartialAnnouncementTag {
+    /// The full tag
+    pub tag: Option<String>,
+    /// The version we're announcing (if doing a unified version announcement)
+    pub version: Option<Version>,
+    /// The package we're announcing (if doing a single-package announcement)
+    pub package: Option<PackageIdx>,
+    /// whether we're prereleasing
+    pub prerelease: bool,
+}
+
 /// Parse the announcement tag and determine what we're announcing
-pub(crate) fn parse_tag(
+///
+/// `tag` being None here is equivalent to `--tag` not being passed, and tells us to infer
+/// the tag based on things like "every package has the same version, assume we're
+/// announcing that version".
+///
+/// `needs_coherent_announcement_tag = false` tells us to produce a result even if inference
+/// fails to find a tag that will unambiguously work. This is used by commands like `init`
+/// and `generate` which want to consider "everything" even if the user never actually
+/// could announce everything at once. In this case dummy values will appear in every field
+/// except for `AnnouncementTag::rust_releases` which will contain every distable binary
+/// in the workspace.
+pub(crate) fn select_tag(
     graph: &DistGraphBuilder,
```

Comment:
this is starting to feel like a struct and methods

Reply:
Yep, planned.